Automattic/jetpack
# Boost: add cache rebuild functionality

This PR adds the cache rebuild functionality from WP Super Cache to the cache in Boost. Instead of deleting expired files, they will be renamed, and if that page is requested, the cache file will be renamed back to its original name. The first visitor won't use that cache file, they'll generate a new one that overwrites the old one. Other visitors seeing the page will use that old cache page, *until* the first visitor generates a cache page.

There's a longer explanation in the issue #37054.

This PR differs from the description in the issue in one way. If the "rebuild file" is older than the cache TTL, the cache file is deleted and not used.
If a visitor resets a rebuild file and renames it back to the old cache file, the header sent will change from hit or miss to rebuild.

This feature is enabled by default in WPSC when site owners enable caching on the "easy" settings page, and this will be enabled by default in Boost as well. There is no UI to disable it.

The `rename()` commands added have to be silenced because two visitors may hit the site at the same time and go through the same code, but one will rename the file before the other.

Should I add a constant to disable rebuild mode?

Fixes #37054

## Proposed changes:
* When attempting to serve a cached page, the system will look for a "rebuild cache file", and rename it back to its original name.
* If the file to be renamed is too old, it is deleted.
* The rebuild cache file isn't used by the first visitor
* Anywhere where a cache file is deleted, apart from uninstalling, the file is rebuilt instead.
* The `delete_directory` function has been renamed `walk_directory` because it handles rebuilding files too.
* The recursive `delete_expired_files` function has been renamed to `gc_expired_files` and has an extra parameter, `$action` that tells it to delete or rebuild files.
* Added `rebuild_file` and `restore_file` functions to handle the renaming.
* Added `reset_rebuild_file` function to check if rebuild file is there, the age, and renaming it back.
* The `garbage_collect` function now calls `delete_expired_files` with the action set to REBUILD.
* The `invalidate` function was updated to support rebuilding.
* The "action", rebuild or delete, must be passed to various functions because the plugin needs to delete the cache files sometimes, like if a post is deleted.

### Other information:


## Jetpack product discussion
pc9hqz-2QH-p2

## Does this pull request change what data or activity we track or use?
no

## Testing instructions:
* Install this PR
* Enable the Cache module and enable logging.
* As a logged out visitor, go to one of the blog posts to generate a cache file.
* Bring up the webdev network console, and filter out everything but html. Confirm you can see the "X-Jetpack-Boost-Cache: hit" header.
* Edit that page in another window and update it.
* Reload the page in the logged out window and verify the "X-Jetpack-Boost-Cache" header value is now "rebuild".
* You can also check the cache directory after you've edited the page as the file should have the extension ".rebuild.html". After viewing the page in a logged out window, the file will be updated with the new cache file.
* You could probably test with multiple clients by using a shell script to launch multiple `wget` or `curl` commands with "&" to put them in the background.
* Make a post private, and make sure the post can't be seen by a logged out browser.
* Delete a comment and make sure the post cache file is deleted.
